### PR TITLE
Refactor of notification system

### PR DESCRIPTION
**Pull request by sbryzak**
This PR is a substantial refactoring of the notifications system.  It introduces a more flexible NotificationBuilder API which allows the new notification context to be populated at creation time, inversing the current design of populating the context at notification delivery time.

e2e tests: https://github.com/codeready-toolchain/toolchain-e2e/pull/362

Fixes https://issues.redhat.com/browse/CRT-821

### CODE REVIEW

File: controllers/deactivation/deactivation_controller_test.go [hunk 1]
```diff
@@ -376,7 +377,7 @@ func prepareReconcile(t *testing.T, name string, initObjs ...runtime.Object) (re
 
 func newObjectMeta(name, email string) metav1.ObjectMeta {
 	if name == "" {
-		name = uuid.NewV4().String()
+		name = uuid.Must(uuid.NewV4()).String()
```
[the remark this answers is not in the document]
Comment (sbryzak) [reply controllers/deactivation/deactivation_controller_test.go:380]: We were supposed to make this change months ago, I'm not sure whether it was missed here or if it somehow crept back into the codebase.  The satori lib is unmaintained and has flaws (see [here](https://github.com/gofrs/uuid)) while the gofrs lib is actively maintained.

File: test/usersignup.go [hunk 2]
```diff
@@ -7,11 +7,11 @@ import (
 
 	"github.com/codeready-toolchain/toolchain-common/pkg/condition"
 	"github.com/codeready-toolchain/toolchain-common/pkg/states"
+	"github.com/gofrs/uuid"
 
 	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
 	"github.com/codeready-toolchain/toolchain-common/pkg/test"
 
-	uuid "github.com/satori/go.uuid"
```
Comment (MatousJobanek) [inline test/usersignup.go:14]: it would be much better to do all these changes in a separate PR

**Review by MatousJobanek — COMMENTED** [review]
Looks really good - I like the new fluent API that you introduced with the builder :1st_place_medal: :+1: 

I've noticed that you didn't modify the logic that creates the notification for sandbox admins when ToolchainStatus is not ready for some time: https://github.com/codeready-toolchain/host-operator/blob/master/controllers/toolchainstatus/toolchainstatus_controller.go#L390-L410
The same is also in the case of the "provisioned notification" https://github.com/codeready-toolchain/host-operator/blob/master/controllers/masteruserrecord/sync.go#L196-L218 - it still uses the old way

**Comment by sbryzak** [comment]
@MatousJobanek thanks for the great review, I'll get to making those changes you suggested however I've been fighting with the e2e tests yet again for the last couple of days and am still unable to get them running.  I'll hopefully have all the issues with them sorted out soon after which I'll address your comments here and add the updated e2e tests for the notification overhaul.

File: controllers/notification/notification_controller.go [hunk 3]
```diff
@@ -81,25 +81,13 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
...
-
 	// if the environment is set to e2e do not attempt sending via mailgun
 	if r.Config.GetEnvironment() != "e2e-tests" {
 		// Send the notification via the configured delivery service
-		err = r.deliveryService.Send(notCtx, notification)
+		err = r.deliveryService.Send(notification)
 		if err != nil {
 			reqLogger.Error(err, "delivery service failed to send notification",
-				notCtx.KeysAndValues()...,
+				notification.Spec.Context,
```
[the remark this answers is not in the document]
Comment (sbryzak) [reply controllers/notification/notification_controller.go:90]: Fixed.

File: controllers/notification/notification_controller_test.go [hunk 4]
```diff
@@ -30,10 +35,14 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
+const (
+	operatorNamespace = "toolchain-host-operator"
+)
```
[the remark this answers is not in the document]
Comment (sbryzak) [reply controllers/notification/notification_controller_test.go:40]: I'm not sure, this may have been legacy.  I've updated it to use `test.HostOperatorNs` anyway.

File: controllers/notification/notification_builder.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	b.options = append(b.options, func(n *toolchainv1alpha1.Notification) error {
+		for k, v := range keysAndValues {
+			n.Spec.Context[k] = v
+		}
+		return nil
+	})
+	return b
+}
+
+func (b *notificationBuilderImpl) WithUserContext(userID string) Builder {
```
[the remark this answers is not in the document]
Comment (sbryzak) [reply controllers/notification/notification_builder.go:126]: That makes it a little cleaner, good idea.

File: controllers/notification/notification_controller_test.go [hunk 6]
```diff
@@ -420,9 +401,34 @@ func newController(t *testing.T, notification *toolchainv1alpha1.Notification, d
...
-	request := reconcile.Request{
+
+	return controller, cl
+}
+
+func reconcileNotification(ctx context.Context, reconciler *Reconciler, notification *toolchainv1alpha1.Notification) (ctrl.Result, error) {
+	return reconciler.Reconcile(ctx, reconcile.Request{
 		NamespacedName: test.NamespacedName(test.HostOperatorNs, notification.Name),
+	})
+}
```
[the remark this answers is not in the document]
Comment (sbryzak) [reply controllers/notification/notification_controller_test.go:412]: Good suggestion, changed.

File: controllers/usersignup/usersignup_controller.go [hunk 7]
```diff
@@ -8,6 +8,8 @@ import (
 	"strconv"
 	"strings"
 
+	notpkg "github.com/codeready-toolchain/host-operator/controllers/notification"
```
[the remark this answers is not in the document]
Comment (sbryzak) [reply controllers/usersignup/usersignup_controller.go:11]: Nice suggestion, changed.

File: test/usersignup.go (hunk 2 above)
Comment (sbryzak) [reply test/usersignup.go:14]: We had previously decided to update this dependency any time we came across it, there is no impact on the main focus of this PR so no harm in including the update here.

**Comment by sbryzak** [comment]
> @sbryzak I've noticed that some functions of the `notification_builder.go` are not covered by tests - you can see that from the code-coverage report. could you please add a test for that file as well?

It's strange that the coverage didn't pick this up, I'm pretty sure most if not all of the functions should have been exercised indirectly via other unit tests.  Anyway I've added a bunch of direct tests.
